Introduce user scanner to Ruler/Alertmanager

SungJin1212 opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with master
-->

This PR introduces a user scanner to reduce the number of list calls to object storage.

**Which issue(s) this PR fixes**:
Fixes #6914 

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

SungJin1212 commented:
@yeya24 
Yeah, agree. What about `/pkg/userscanner`?

yeya24 commented:
>  Yeah, agree. What about /pkg/usersscanner?

I am thinking about `/pkg/util/users` which includes everything user related, not just user scanner.

SungJin1212 commented:
> I am thinking about /pkg/util/users which includes everything user related, not just user scanner.
ok, it looks good.
I moved user-related files to `/pkg/util/users` and renamed these files.
`users_scanner_config.go` -> `scanner_config.go`
`updater.go` -> `index_updater.go`

danielblando commented:
What is the difference for us between user and tenant? It is a bit weird to have `/pkg/util/users/tenant/`.

SungJin1212 commented:
@danielblando 
I think `tenant` should be integrated into `users`, but I've left it like this for now due to a cyclic import. I'll look into a way to integrate it.

SungJin1212 commented:
@danielblando @yeya24 
I've moved the `tenant` package to the `users` package. I also deprecated `util.StringsContain` and replaced it with `slices.Contains` to resolve a cyclic import.

yeya24 commented on pkg/util/errors/errors.go:
```diff
@@ -50,3 +54,14 @@ func ErrorIs(err error, f func(err error) bool) bool {
 	}
 }
+
+func IsOneOfTheExpectedErrors(f ...objstore.IsOpFailureExpectedFunc) objstore.IsOpFailureExpectedFunc {
```
Why this function is not part of the bucket package but in a generate `pkg/util/errors`? This seems specific to object store failures so I am not sure if it should be in the generic error file

SungJin1212 replied:
I return it to its original location, as I had temporarily moved it to resolve a cyclic import.

SungJin1212 replied:
I moved `IsOneOfTheExpectedErrors` to the `bucket` package. It causes a cyclic import if it is in the `tsdb` package.

SungJin1212 replied:
What do you think of using only the `list` strategy and only exposing `cache_ttl` as config?

yeya24 commented on pkg/alertmanager/alertstore/bucketclient/bucket_client.go:
```diff
@@ -45,27 +48,36 @@ type BucketAlertStore struct {
+func NewBucketAlertStore(bkt objstore.InstrumentedBucket, userScannerCfg users.UsersScannerConfig, cfgProvider bucket.TenantConfigProvider, logger log.Logger, reg prometheus.Registerer) (*BucketAlertStore, error) {
+	alertBucket := bucket.NewPrefixedBucketClient(bkt, alertsPrefix)
+
+	usersScanner, err := users.NewScanner(userScannerCfg, alertBucket, logger, extprom.WrapRegistererWith(prometheus.Labels{"component": "alertmanager"}, reg))
```
Do we support user index based user scanner? Who will work as the user index updater for AM and Ruler bucket?

yeya24 replied:
Yeah, for compactor, we use Ring to choose one compactor pod to periodically update the user index in the storage bucket. I think we need the same thing for Ruler/AM as they are using different S3 buckets

SungJin1212 replied:
How about adding a new config `index_cleanup_interval` for `userIndexUpdater` and making `userIndexUpdater` update the user index file in its own loop?
It would be good to make the `userIndexUpdater` sub-service of the Compactor, AM, and the Ruler.

yeya24 commented on pkg/storage/tsdb/bucketindex/loader_test.go:
```diff
@@ -17,16 +17,15 @@ import (
-	cortex_testutil "github.com/cortexproject/cortex/pkg/storage/tsdb/testutil"
 	"github.com/cortexproject/cortex/pkg/util/services"
 	"github.com/cortexproject/cortex/pkg/util/test"
+	cotex_testutil "github.com/cortexproject/cortex/pkg/util/testutil"
```
This seems a typo.

yeya24 commented on docs/blocks-storage/store-gateway.md:
```diff
@@ -2017,6 +2017,11 @@ blocks_storage:
 
+    # How frequently user index file is updated, it only take effect when user
+    # scan stratehy is user_index.
+    # CLI flag: -blocks-storage.users-scanner.user-index.cleanup-interval
```
It only takes effect when user scan strategy
